### PR TITLE
Brandon dev: dural lesion modifications

### LINKED ISSUE #68: strange behavior of mass effect in dural lesions
Mass effect is meant to push brain tissue away from the growing lesion, but based on its current impementation this can lead to some irregular behavior such as the lesion pulling tissue towards it, see this example of a lesion exerting more mass_effect:

381228547-224f9f4b-45db-4733-b453-34f0fbe4d9c7

I recommend playing around with lesion creation and the different parameters here in the [interactive lesion playground notebook](https://github.com/DIDSR/PedSilicoICH/blob/brandon_dev/notebooks/lesion_playground.ipynb)

Because of this issue, the commits for modifying mass effect strength are in a pull request: https://github.com/DIDSR/PedSilicoICH/pull/66

From the brandon_dev branch, see in particular:

https://github.com/DIDSR/PedSilicoICH/blob/brandon_dev/src/pedsilicoICH/lesion_definition.py#L141C1-L154C79

### PR DESCRIPTION
**Pull request by brandonjnelsonFDA**
This PR addresses the following issues (it is also an example of how NOT to do a PR as it is multi-part and too big, see [PR best practices](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/best-practices-for-pull-requests))

1.  [adding a random seed]: #64 for each study to the metadata so each simulation can be reproduced later (including whole datasets) see CT_dataset_pipeline.py
2. [dural mass effect unexpected behaviors]: I noticed the warped mass effect image from [insert_dural_3D](https://github.com/brandonjnelsonFDA/PedSilicoICH/blob/97e98a3a4ca50c85bbbf9e60ed338576e54e2d4b/src/pedsilicoICH/ground_truth_definition/phantoms.py#L458) wasn't being used so I renamed from volume --> [HU_volume](https://github.com/brandonjnelsonFDA/PedSilicoICH/blob/0f18fa6d6f96dc77d4ae1af551ab5df0615e177a/src/pedsilicoICH/ground_truth_definition/phantoms.py#L448)
- this could have been done on purpose due to mass effect errors I added the following correction to replace bright splotches from bone voxels warped into the brain with mean brain values see src/pedsilicoICH/lesion_definition.py lines 305-307.

code to reproduce:
```
seed = 42
age = 6.5
phantom = load_phantom(age=age)
phantom.insert_lesion('epidural', volume=40, mass_effect=0.8, seed=seed)
ctshow(phantom._phantom[phantom._lesion_coords[0][0]])
```

image
Epidural lesion, no mass effect

image
Epidural lesion before applying the correction, lesion_definition.py lines 305-307

image
Epidural lesion post correction, while not perfect, blurring from the CT acquisition will smooth our remaining rough transitions 

-  Note, this is just a band-aid, and I notice the current warping can be a little eratic #68, see for example the following gif example of increasing the amount of mass effect (see change 3):
Stack

For small amounts of mass effect, <0.3 in [0, 1], it acts as expected pushing tissue away, but for larger values tissue gets pulled in which is not the desired effect

3. [controlling mass effect strength for dural lesions]: similar to [add_round_lesion](https://github.com/brandonjnelsonFDA/PedSilicoICH/blob/97e98a3a4ca50c85bbbf9e60ed338576e54e2d4b/src/pedsilicoICH/ground_truth_definition/phantoms.py#L359), mass_effect can be entered either as a boolean or a float between 0, 1 to control strength, where 0 is no mass effect, and 1 is full strength. This is done in a similar manner by controlling the number of morphological erosions between the [warp src and dst](https://github.com/brandonjnelsonFDA/PedSilicoICH/blob/97e98a3a4ca50c85bbbf9e60ed338576e54e2d4b/src/pedsilicoICH/ground_truth_definition/phantoms.py#L359), I found this to work for dural lesions too because the skull map acts as a constraint

4. [move file paths to environment variables]: Rather than hard coding paths, adds python-dotenv to load_phantoms so that files can be run on openhpc, luna, and other user systems without updating hard coded file paths

WARNING: this is a breaking change, you need to add a file called '.env' with the line:

PHANTOM_DIRECTORY=/gpfs_projects/brandon.nelson/pedsilicoICH/phantoms

Where PHANTOM_DIRECTORY can be a users phantom location. this needs to be manually added to use MIDA phantoms as .env is on the .gitignore list

See load_phantom for more details

### CODE REVIEW

**Comment by brandonjnelsonFDA** [comment]
@jmweaver-FDA raised a good point that removing the try except clause removed the ability to search for suitable init_slices, this was addressed in commit f5975f8227fcd9ae6627585f93f6b1baf63f8048 in this PR by only randomly choosing an init slice from the list of candidate slices that can fit a lesion of the desired volume, see lines 85-92 of [lesion_definition](https://github.com/DIDSR/PedSilicoICH/blob/brandon_dev/src/pedsilicoICH/lesion_definition.py)

**Comment by jmweaver-FDA** [comment]
Adding my comments in chunks as I go through the commits due to length:

I like the changes in commits [a3ed83b](https://github.com/DIDSR/PedSilicoICH/pull/66/commits/a3ed83b946195c1554b511496b03bd4f4a5c5123) and [f5975f8](https://github.com/DIDSR/PedSilicoICH/pull/66/commits/f5975f8227fcd9ae6627585f93f6b1baf63f8048) to clean up and potentially speed up the selection of initial slice for the subdural and epidural lesions. Using the expected area of the lesion (AB from the ABC/2) lesion and the size of the slice from the skull map was a smart way to find a large enough slice without iteratively searching. Once I have this PR pushed to Luna, we'll be able to see if any subdural/epidural lesions fail to insert when generating larger datasets.

**Comment by jmweaver-FDA** [comment]
I can start to see the reason behind the issue in #68 - it's due to the implementation of mass_effect strength similar to the one used for the round hemorrhages.

In short, due to its current implementation, the method of eroding the perimeter of the filled array cannot work for subdural and epidural lesions. This is because the `boundary_coords` connecting the start and end point along the skull/dura boundary should never be changed and these voxels should never move. The `src` list should only contain voxels along this boundary, and `dst` should contain their counterparts (sampled from `connect_coords`) to warp to. I think when `strength` increases the order of `src` and `dst` is somehow getting out of order leading to the unexpected warping.

The most logical way to support some kind of mass_effect strength for the dural hemorrhages (that I can think of now) is to create a weaker Bezier curve (by reducing either `bezier_weight` or changing `bezier_middle`) and then using the coordinates of this weaker `connect` coords to warp the brain mask.

Eventually, if we switch to a region growing approach it might be much easier to have a similar mass effect approach as the round hemorrhages, but for now I think we should revert the mass effect changes in `insert_dural_3D`. However, I think it would be worth moving them to a different branch/stash for posterity in case we try the method again.

File: src/pedsilicoICH/lesion_definition.py
Comment (brandonjnelsonFDA) [inline src/pedsilicoICH/lesion_definition.py:1]: After reverting back to 6b6af70ea3d6217560f1932242724f86a93f8655, the only changes to this file are modest. @jmweaver-FDA

**Pull request merged**